test/GasAccounting.t.sol

```diff
@@ -0,0 +1,318 @@
+    }
+
+    function test_validateBalances() public {
+        assertFalse(mockGasAccounting.validateBalances());
+
+        mockGasAccounting.trySolverLock(solverOp);
+        mockGasAccounting.reconcile{ value: initialClaims }(executionEnvironment, solverOp.from, 0);
+
+        assertTrue(mockGasAccounting.validateBalances());
+    }
+
+    function test_contribute() public {
+        vm.expectRevert(
+            abi.encodeWithSelector(FastLaneErrorsEvents.InvalidExecutionEnvironment.selector, executionEnvironment)
+        );
+        mockGasAccounting.contribute();
+
+        uint256 contributeValue = 1000;
+        deal(executionEnvironment, contributeValue);
+
+        vm.prank(executionEnvironment);
+        mockGasAccounting.contribute{ value: contributeValue }();
+
+        assertEq(mockGasAccounting.deposits(), contributeValue);
+    }
+
+    function test_borrow() public {
+        uint256 borrowedAmount = 5000;
+
+        vm.expectRevert(
+            abi.encodeWithSelector(FastLaneErrorsEvents.InvalidExecutionEnvironment.selector, executionEnvironment)
+        );
+        mockGasAccounting.borrow(borrowedAmount);
+
+        vm.prank(executionEnvironment);
+        vm.expectRevert(
+            abi.encodeWithSelector(FastLaneErrorsEvents.InsufficientAtlETHBalance.selector, 0, borrowedAmount)
+        );
+        mockGasAccounting.borrow(borrowedAmount);
+
+        deal(
+            address(mockGasAccounting),
+            initialClaims + borrowedAmount // claims + amount borrowed
+        );
+
+        vm.prank(executionEnvironment);
+        mockGasAccounting.borrow(borrowedAmount);
+
+        assertEq(executionEnvironment.balance, borrowedAmount);
+    }
+
+    function test_shortfall() public {
+        assertEq(mockGasAccounting.shortfall(), initialClaims);
+
+        deal(executionEnvironment, initialClaims);
+        vm.prank(executionEnvironment);
+        mockGasAccounting.contribute{ value: initialClaims }();
+
+        assertEq(mockGasAccounting.shortfall(), 0);
+    }
+
+    function test_reconcile() public {
+        vm.expectRevert(
+            abi.encodeWithSelector(FastLaneErrorsEvents.InvalidExecutionEnvironment.selector, executionEnvironment)
+        );
+        mockGasAccounting.reconcile(makeAddr("wrongExecutionEnvironment"), solverOp.from, 0);
+
+        mockGasAccounting.trySolverLock(solverOp);
+        vm.expectRevert(abi.encodeWithSelector(FastLaneErrorsEvents.InvalidSolverFrom.selector, solverOp.from));
+        mockGasAccounting.reconcile(executionEnvironment, makeAddr("wrongSolver"), 0);
+
+        vm.expectRevert(abi.encodeWithSelector(FastLaneErrorsEvents.InsufficientTotalBalance.selector, initialClaims));
+        mockGasAccounting.reconcile(executionEnvironment, solverOp.from, 0);
+
+        assertTrue(mockGasAccounting.reconcile{ value: initialClaims }(executionEnvironment, solverOp.from, 0));
+    }
+
+    function test_assign() public {
```

Comment:
I haven't been through this logic in detail, it's gonna take me a while to go through and understand the accounting logic here. I will try set aside some time for that soon.

But I think we should also be testing boundary conditions on this, e.g. `amount` gets cast from a unit256 to a uint128:

```
            uint128 amt = uint128(amount);
```

We should probably test `amount` with inputs > uint128.max to see how it behaves.

Reply:
Added a test for amounts > type(uint128).max. It overflows and reverts (using Solidity 0.8.x) which is expected. We shouldn't have to worry about these values being passed into the function as they should always come from `msg.value` or a `gasUsed` amount, and those should never exceed type(uint128).max as that is more ETH than will probably ever exist.

We do need to be careful about deploying this on a chain where quantities of the native token could reach type(uint128).max.